pkg/job/scraper.go

```diff
@@ -0,0 +1,241 @@
+package job
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/nerdswords/yet-another-cloudwatch-exporter/pkg/job/cloudwatchrunner"
+
+	"github.com/nerdswords/yet-another-cloudwatch-exporter/pkg/clients/account"
+	"github.com/nerdswords/yet-another-cloudwatch-exporter/pkg/logging"
+	"github.com/nerdswords/yet-another-cloudwatch-exporter/pkg/model"
+)
+
+type Scraper struct {
+	jobsCfg       model.JobsConfig
+	logger        logging.Logger
+	runnerFactory runnerFactory
+}
+
+type runnerFactory interface {
+	GetAccountClient(region string, role model.Role) account.Client
+	NewResourceMetadataRunner(logger logging.Logger, region string, role model.Role) ResourceMetadataRunner
+	NewCloudWatchRunner(logger logging.Logger, region string, role model.Role, job cloudwatchrunner.Job) CloudwatchRunner
+}
+
+type ResourceMetadataRunner interface {
+	Run(ctx context.Context, region string, job model.DiscoveryJob) ([]*model.TaggedResource, error)
+}
+
+type CloudwatchRunner interface {
+	Run(ctx context.Context) ([]*model.CloudwatchData, error)
+}
+
+func NewScraper(logger logging.Logger,
+	jobsCfg model.JobsConfig,
+	runnerFactory runnerFactory,
+) *Scraper {
+	return &Scraper{
+		runnerFactory: runnerFactory,
+		logger:        logger,
+		jobsCfg:       jobsCfg,
+	}
+}
+
+func (s Scraper) Scrape(ctx context.Context) ([]model.TaggedResourceResult, []model.CloudwatchMetricResult, []Error) {
+	// Setup so we only do one GetAccount call per region + role combo when running jobs
+	roleRegionToAccount := map[model.Role]map[string]func() (string, error){}
+	jobConfigVisitor(s.jobsCfg, func(_ any, role model.Role, region string) {
+		if _, exists := roleRegionToAccount[role]; !exists {
+			roleRegionToAccount[role] = map[string]func() (string, error){}
+		}
+		roleRegionToAccount[role][region] = sync.OnceValues[string, error](func() (string, error) {
+			accountID, err := s.runnerFactory.GetAccountClient(region, role).GetAccount(ctx)
+			if err != nil {
+				return "", fmt.Errorf("failed to get Account: %w", err)
+			}
+			return accountID, nil
+		})
+	})
+
+	var wg sync.WaitGroup
+	mux := &sync.Mutex{}
+	jobErrors := make([]Error, 0)
+	metricResults := make([]model.CloudwatchMetricResult, 0)
+	resourceResults := make([]model.TaggedResourceResult, 0)
+	s.logger.Debug("Starting job runs")
+
+	jobConfigVisitor(s.jobsCfg, func(job any, role model.Role, region string) {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			var namespace string
+			jobAction(s.logger, job, func(job model.DiscoveryJob) {
+				namespace = job.Type
+			}, func(job model.CustomNamespaceJob) {
+				namespace = job.Namespace
+			})
```

Comment:
Wondering if it's time to go and rename `job.Type` instead of doing this? (it's an internal model after all).

Reply:
More generally, maybe instead of `jobAction` we can use a minimal interface through the models you created in `cloudwatchrunner.DiscoveryJob` and `cloudwatchrunner.CustomNamespaceJob`?